### DISCUSSION TITLE
Inheritance & Generics

### DISCUSSION THREAD

**Discussion by gilles-mnr** [wasm-bindgen/wasm-bindgen#4087] [Q&A] [answered]
### Summary
Hello, i'm creating a game engine and I'm trying to implement collisions. I need to  implement it for all `Graphic` graphic is the parent of many JavaScript classes that extends it (`Line`, `Circle`, `Ellipse`, `Rectangle`). In my Rust code I need to do a function that compares to `Graphic` instance (that may be a child) and i need to export that function to JavaScript

### Additional Details
```js
export class Graphic { ... }
export class Rectangle extends Graphic  { ... } 
...
```

```rust
#[wasm_bindgen]
extern "C" {
    pub type Graphic;
    pub type Circle;
    pub type Ellipse;
    pub type Line;
    pub type Rectangle;
}
```

So firstly i would have to know what kind of instance it is so here is just a function to do that. Unfortunately it doesn't work: the output is always "???"

```rust 
#[wasm_bindgen]
pub fn show_info(graphic: &Graphic) {
    if graphic.is_instance_of::<Ellipse>() {
        console_log!("Ellipse");
    } else if graphic.is_instance_of::<Circle>() {
        console_log!("Circle");
    } else if graphic.is_instance_of::<Rectangle>() {
        console_log!("Rectangle");
    } else if graphic.is_instance_of::<Line>() {
        console_log!("Line");
    } else {
        console_log!("???");
    }
}
```

```js
let testRect = new cazan.graphics.Rectangle({
    game: game,
    position: {x: 10, y: 10},
    dimensions: {width: 50, height: 50},
    styles: {
        graphic: blackStyle
    }
})

show_info(testRect)
```

Thanks to you all for your help :). If you need more details i'm answering for sure

**Comment by daxpedda** [comment]
I'm unfortunately unfamiliar with integrating with JS, but everything here looks right.
We also have a test on this: https://github.com/rustwasm/wasm-bindgen/blob/0cbca3174acad5111725f7e7d4f1ab7c10bebaaf/tests/wasm/jscast.rs.

I don't think `#[wasm_bindgen(extends = Graphic)]` is required here, but feel free to try it, it might also just be useful for you as well.

Could you provide a complete but minimal example that reproduces your issue?

**Reply by gilles-mnr** [reply]
I supposed that putting "Graphic" as the type of the argument could "limit" the variable so I tried with "JsValue" but it still doesn't work.
I added the `#[wasm_bindgen(extends = Graphic)]` above the types (not Graphic).
What do you miss to be able to reproduce my issue ?

**Reply by daxpedda** [reply]
The JS side I guess.
But I would like a complete example that I can just look at instead of having to make one myself with limited information!

In general, giving a complete minimal reproducible example is the way to go about those sort of issues.

**Reply by gilles-mnr** [reply]
We use rollup so all the js is in the same file "cazan.js" 

`lib.rs`
```rs
use wasm_bindgen::prelude::*;

#[wasm_bindgen]
extern "C" {
    pub type Graphic;
    #[wasm_bindgen(extends = Graphic)]
    pub type Circle;
    #[wasm_bindgen(extends = Graphic)]
    pub type Ellipse;
    #[wasm_bindgen(extends = Graphic)]
    pub type Line;
    #[wasm_bindgen(extends = Graphic)]
    pub type Rectangle;
}

#[wasm_bindgen(start)]
pub fn main() {
    utils::set_panic_hook(); // Generated
}

#[wasm_bindgen]
extern "C" {
    #[wasm_bindgen(js_namespace = console)]
    pub fn log(s: &str);
}


#[wasm_bindgen]
pub fn show_info(graphic: &JsValue) {

    if graphic.has_type::<Rectangle>() {
        log("rectangle");
    } else if graphic.has_type::<Graphic>() {
        log("not rectangle");
    } else {
        log("...")
    }
}
```
`cazan.js`
```js
export class Graphic {}
export class Ellipse extends Graphic {}
export class Circle extends Graphic {}
export class Rectangle extends Graphic {}
export class Line extends Graphic {}
```

`app.js`
```js
import * as cazan from "/cazan.js"
import * as wasm from "./pkg/wasm.js" // The WASM

let testRect = new cazan.Rectangle()
wasm.show_info(testRect) // "..."
```
```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <script src="/app.js" type="module"></script>
</head>
<body>
</body>
</html>
```
Note: 
i found a trick that can give me the "constructor name" of the js object but i would still like to do it with the clean way:
```rs
pub fn get_constructor_name(obj: &JsValue) -> Option<String> {
    if let Some(obj_ref) = obj.dyn_ref::<Object>() {
        if let Ok(constructor) = Reflect::get(obj_ref, &"constructor".into()) {
            if let Some(constructor_name) = constructor.dyn_ref::<js_sys::Function>() {
                return constructor_name.name().as_string();
            }
        }
    }
    None
}
```

**Reply by daxpedda** [reply, marked as answer]
```rust
#[wasm_bindgen(raw_module = "./cazan.js")]
extern "C" {
    pub type Graphic;
    #[wasm_bindgen(extends = Graphic)]
    pub type Circle;
    #[wasm_bindgen(extends = Graphic)]
    pub type Ellipse;
    #[wasm_bindgen(extends = Graphic)]
    pub type Line;
    #[wasm_bindgen(extends = Graphic)]
    pub type Rectangle;
}
```
Does the trick for me.
See the [`wasm-bindgen` Guide on this topic](https://rustwasm.github.io/docs/wasm-bindgen/reference/attributes/on-js-imports/raw_module.html).

**Reply by gilles-mnr** [reply]
It works!

Cazan.js is in the parent directory so i used `../cazan.js` and the graphics were under `graphics` so i used ´js_namespace = graphics`.

Actually i have tried with `module` but it didn't work because it's not in the same directory I guess. I didn't know that `raw_module` exists;

Thank you very much :)
